### PR TITLE
Support private Route53 zones

### PR DESCRIPTION
**Pull request by JonRoma**
*   Module now supports private Route53 zones. In this case, the A record is set to the private rather than public IP.

*   One can now set "associate_public_ip_address" to false for instances that don't have public Route53 presence.

These discoveries were made in the course of testing networking for a new VPC.

### CODE REVIEW

**Comment by JonRoma** [comment]
@ddriddle, would you please be so kind as to give this a review?

The logic to compute the IP address based on whether an elastic IP address is selected could probably be simplified, particularly in the case of a Route 53 alias in a private zone, where EIP makes no sense. In theory specifying EIP and private_zone should produce an error, but I didn't bother with that.

**Review by cmaturi — DISMISSED** [review]

File: .github/workflows/terraform.yml [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# Some of this logic was stolen from https://www.freecodecamp.org/news/a-lightweight-tool-agnostic-ci-cd-flow-with-github-actions/
+
+name: terraform
+
+on:
+  push:
+#   branches:
```
Comment (ddriddle) [inline .github/workflows/terraform.yml:7]: Please, drop the unnecessary comments on lines 7 and 8.

File: .github/workflows/terraform.yml [hunk 2]
```diff
@@ -0,0 +1,81 @@
+# Some of this logic was stolen from https://www.freecodecamp.org/news/a-lightweight-tool-agnostic-ci-cd-flow-with-github-actions/
```
Comment (ddriddle) [inline .github/workflows/terraform.yml:1]: You have made so many changes I don't think you need to credit this link anymore. IMO.

File: .github/workflows/terraform.yml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        with:
+          fetch-depth: 1
+
+      - name: terraform setup
+        uses: hashicorp/setup-terraform@v1
+        with:
+          terraform_version: 0.12.29
+      #   cli_config_credentials_token: ${{ secrets.TF_API_TOKEN }}
+
+#     TODO: This step duplicates work done by the Makefile.
```
Comment (ddriddle) [inline .github/workflows/terraform.yml:27]: Please, remove this comment block (lines 27-31)

File: .github/workflows/terraform.yml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+#       run: |
+#         terraform fmt -check -recursive
+
+      - name: run make
+      # env:
+      #   TOKEN: ${{ secrets.TOKEN }}
+        run: |
+          make all
+
+#     - name: terraform init
```
Comment (ddriddle) [inline .github/workflows/terraform.yml:39]: Please remove this unused code (lines 39-81).

File: route53.tf [hunk 5]
```diff
@@ -1,15 +1,28 @@
...
+    ttl          = lookup(a, "ttl", 60)
+    }
+  ]
+}
+
 data "aws_route53_zone" "selected" {
   count = length(var.alias)
 
-  name = lookup(var.alias[count.index], "domain", "")
+  name         = local.alias[count.index].domain
```
Comment (ddriddle) [inline route53.tf]: @JonRoma What is the advantage of using the `local.alias` compared to just using a straight `lookup` as was done before?

File: route53.tf [hunk 6]
```diff
@@ -1,15 +1,28 @@
+locals {
+  alias = [for a in var.alias : {
+    domain       = a.domain
+    hostname     = a.hostname
```
Comment (ddriddle) [inline route53.tf]: Before if the hostname was not specified it defaulted to `var.name`. Did you intend to change that behavior? It is documented in the README and the code for that use to be on line 11 below.

Comment (ddriddle) [reply route53.tf]: Please change line 4 to:

```
hostname = lookup(a, "hostname", var.name)
```

**Review by ddriddle — CHANGES_REQUESTED** [review]
@JonRoma I think the code for calculating the private IP and DNS is good, as it allows for the code to evolve in a way that we can be sure that the values returned are always correct. I agree, you could simplify the code but I prefer the code as is since it is more defensive.

I found a bug. The hostname should default to `var.name` if not specified. That logic was dropped. If you intended that change then you need to update the documentation, otherwise please apply the one line change I suggested.

Overall the code looks good, but I wonder if `local.alias` makes the code more complicated. I'm not sure. You may want to consider refactoring it out. There may well be a very good reason for the object that I am not seeing.

File: .github/workflows/terraform.yml (hunk 2 above)
Comment (JonRoma) [reply .github/workflows/terraform.yml:1]: @ddriddle, I should have asked you not to worry about reviewing the GitHub actions. The file here is a minimally-viable version borrowed from other Terraform projects, and as such does not represent anything new to this repo, nor does it represent what I hope to attain as a gold standard for GitHub Actions for Terraform modules.

This is one of the areas that I hope to work on with you in the future after Shibboleth has died down. To that end, I'm going to ignore the comments on the CI pieces, with the thought that we'll address them together soon, add them to SDG's templates, and disperse them broadly across all our repos so that we can stick a fork in Drone.

File: route53.tf (hunk 5 above)
Comment (JonRoma) [reply route53.tf]: The purpose of the added local variable was to facilitate the addition of the VPC id for private zones in a way that is reasonably readable. Perhaps there are other ways to accomplish that – do you have ideas?

While creating a bunch of intermediate variables sometimes annoys me, it's something I had to do heavily in the `enterprise-subnets` module to be generic and to use sophisticated data structures without making the code a complete nightmare of nested maps and lists that no one including me will be able to read three months from now. :-)

I don't think sheer efficiency [by avoiding temporary variables] is worth fretting about in Terraform code, and I tend to err on the side of readability vs. efficiency.

File: route53.tf (hunk 6 above)
Comment (JonRoma) [reply route53.tf]: This was **not** intentional; good catch. I have fixed it in the latest push. 

Please note that, if the user specifies more than one alias but omits the hostname argument on both, Route 53 will reject the `apply` because of a duplicate hostname. That's obviously a user error, however.

**Comment by JonRoma** [comment]
@ddriddle, I have addressed the issue you pointed out, and have commented on the others. If you are satisfied, please give an approving review so that this can go forward. Otherwise, let's talk. :-)

File: route53.tf (hunk 5 above)
Comment (ddriddle) [reply route53.tf]: I'm not worried about efficiency here, my only concern is that the diff is larger than necessary which does make it harder to review and understand why the changes are happening. Other than that your approach is fine. I don't think it improves readability much here, but in general I agree it is a good idea, and much better than spreading the logic all over the place.

> The purpose of the added local variable was to facilitate the addition of the VPC id for private zones in a way that is reasonably readable. Perhaps there are other ways to accomplish that – do you have ideas?

What I would have done is simply inlined the `vpc_id` logic since it is only used in one place. If the logic was needed in more than one place then I would have gone with the same approach you went with. Here is how I would refactor the code:

```terraform
data "aws_route53_zone" "selected" {
  count = length(var.alias)
  
  name         = lookup(var.alias[count.index], "domain", "")
  private_zone = lookup(var.alias[count.index], "private_zone", false)
  vpc_id       = lookup(var.alias[count.index], "private_zone", false) ? data.aws_vpc.selected.id : null
}

resource "aws_route53_record" "default" {
  count = length(var.alias)

  zone_id = element(data.aws_route53_zone.selected.*.zone_id, count.index)
  name    = lookup(var.alias[count.index], "hostname", var.name)
  type    = "A"
  ttl     = lookup(var.alias[count.index], "ttl", 60)
  records = [lookup(var.alias[count.index], "private_zone", false) ? local.private_ip : local.public_ip]
}
```

I'm fine keeping the object or refactoring it out. Your call. Thanks for your comments!

File: route53.tf [hunk 7]
```diff
@@ -1,15 +1,28 @@
...
   count = length(var.alias)
 
-  name = lookup(var.alias[count.index], "domain", "")
+  name         = local.alias[count.index].domain
+  vpc_id       = local.alias[count.index].vpc_id
+  private_zone = local.alias[count.index].private_zone
 }
 
 resource "aws_route53_record" "default" {
   count = length(var.alias)
```
Comment (ddriddle) [inline route53.tf]: I think it would be better to change this to:

```
count = length(local.alias)
```

File: route53.tf [hunk 8]
```diff
@@ -1,15 +1,28 @@
...
+    hostname     = lookup(a, "hostname", var.name)
+    private_zone = lookup(a, "private_zone", false)
+    vpc_id       = lookup(a, "private_zone", false) ? data.aws_vpc.selected.id : null
+    ttl          = lookup(a, "ttl", 60)
+    }
+  ]
+}
+
 data "aws_route53_zone" "selected" {
   count = length(var.alias)
```
Comment (ddriddle) [inline route53.tf:2]: I think it would be better to directly count `local.alias` since `var.alias` is not directly used here anymore:

```
count = length(local.alias)
```

**Review by ddriddle — CHANGES_REQUESTED** [review]
Please change the count lines if you choose not to remove `local.alias`. Thanks for asking for my review and your thoughtful comments. Enjoy the snow :-D

File: route53.tf (hunk 5 above)
Comment (JonRoma) [reply route53.tf]: @ddriddle, honestly this is why I like reviewing each other's code. I added the logic I did because at first glance, the addition of private zone support looked more complicated than it actually was. And when I finished, I was tired enough that I forgot to take a step back and see that the extra data structure wasn't needed.

I've refactored the `local.alias` out of the module.

File: route53.tf (hunk 7 above)
Comment (JonRoma) [reply route53.tf]: This is now moot due to my refactoring, but I agree that the count should match the variable being iterated over, be it in `for_each` or with `count.index`.

As an aside, I found an interesting tidbit in that Terraform's `keys` and `values` function always returns a list that's in lexicographical order, and is hence deterministic. I have used this to my advantage elsewhere. :-)

**Comment by JonRoma** [comment]
@ddriddle, the refactored code is ready for your review.

Also, at a customer request, I added the hostname to the list of template variables passed in to the `user_data` script. I had to derive it from the first element of the `alias` variable. A more logical way would have been to use the value from `as_aws_instance.fqdn`, but that won't work because that value isn't known until the instance is created – a chicken *vs.* egg problem!

File: data.tf [hunk 9]
```diff
@@ -67,12 +67,15 @@ locals {
   efs_file_system_id = lookup(var.efs_file_system, "file_system_id", "")
   efs_mount_point    = lookup(var.efs_file_system, "mount_point", "")
   efs_source_path    = lookup(var.efs_file_system, "source_path", "")
+  hostname           = length(var.alias) > 0 ? format("%s.%s", var.alias[0].hostname, var.alias[0].domain) : null
```
Comment (ddriddle) [inline data.tf]: Line 70 has a bug: hostname is optional, and should default to `var.name` per the docs, but here it is treated like a required variable, so it should be:

```terraform
length(var.alias) > 0 ? format("%s.%s", lookup(var.alias[0], "hostname", var.name), var.alias[0].domain) : null
```

File: route53.tf [hunk 10]
```diff
@@ -1,15 +1,17 @@
...
+  vpc_id       = lookup(var.alias[count.index], "private_zone", false) ? data.aws_vpc.selected.id : null
+  private_zone = lookup(var.alias[count.index], "private_zone", false)
 }
 
 resource "aws_route53_record" "default" {
   count = length(var.alias)
 
   zone_id = element(data.aws_route53_zone.selected.*.zone_id, count.index)
-  name    = lookup(var.alias[count.index], "hostname", var.name)
+  name    = var.alias[count.index].hostname
```
Comment (ddriddle) [inline route53.tf]: Please revert this change, otherwise update the documentation to indicate that `var.name` is not the default.

**Review by ddriddle — CHANGES_REQUESTED** [review]
The documentation still indicates that `var.name` is the default value for hostname in an alias block. The code in this pull request does not implement that feature. I made suggestions in the comments below on how to easily add that feature back. If the feature is no longer desirable then please update the documentation.

File: data.tf (hunk 9 above)
Comment (JonRoma) [reply data.tf]: Code has been refactored per your suggestions. 

To be completely honest, a user omitting something as critical as a hostname or domain from the `alias` block is probably being foolish.

**Comment by JonRoma** [comment]
@ddriddle, please take another look.

**Review by ddriddle — APPROVED** [review]
Everything looks good!

**Review by cmaturi — APPROVED** [review]
